File: code/0005. Longest Palindromic Substring/solution2.java
class Solution {
    private String longest = "";
    
    private void expand(String s, int i, int j) {
        int n = s.length();
        
        while (i > 0 && j < n - 1 && s.charAt(i - 1) == s.charAt(j + 1)) {
            i--;
            j++;
        }
        
        int len = j - i + 1;
        if (len > longest.length()) {
            longest = s.substring(i, j + 1);
        }
    }
    
    public String longestPalindrome(String s) {
        int n = s.length();
        for (int i = 0; i < n; i++) {
            expand(s, i, i);
            if (i < n - 1 && s.charAt(i) == s.charAt(i + 1)) {
                expand(s, i, i + 1);
            }
        }
        
        return longest;
    }
}

File: code/0005. Longest Palindromic Substring/solution2.cpp
class Solution {
private:
    void expand(string& s, int i, int j, string& longest) {
        int n = s.length();
        while (i > 0 && j < n - 1 && s[i - 1] == s[j + 1]) {
            i--;
            j++;
        }
        
        int len = j - i + 1;
        if (len > longest.length()) {
            longest = s.substr(i, len);
        }
    }
    
public:
    string longestPalindrome(string s) {
        string longest = "";
        int n = s.length();
        
        for (int i = 0; i < n; i++) {
            expand(s, i, i, longest);
            if (i < n - 1 && s[i] == s[i + 1]) {
                expand(s, i, i + 1, longest);
            }
        }
        
        return longest;
    }
};